File: paradigms.py

# Imperative

numbers = [1, 2, 3, 4, 5, 6]
result = []

for num in numbers:
    if num % 2 == 0:
        square = num * num
        result.append(square)

print("Imperative:", result)  


# Declarative approach 

numbers = [1, 2, 3, 4, 5, 6]
result = list(map(lambda x: x * x, filter(lambda x: x % 2 == 0, numbers)))

print("Declarative:", result)  




# Functional approach 

from typing import List

def is_even(n: int) -> bool:
    return n % 2 == 0

def square(n: int) -> int:
    return n * n

def process(numbers: List[int]) -> List[int]:
    return list(map(square, filter(is_even, numbers)))



numbers = [1, 2, 3, 4, 5, 6]
result = process(numbers)

print("Functional:", result) 

File: paradigms.go
package main

// 1. Imperative  (in main function)
func processImperative(nums []int) []int {
	var result []int
	for _, n := range nums {
		if n%2 == 0 {
			result = append(result, n*n)
		}
	}
	return result
}

// 2. Declarative

func filter(nums []int, test func(int) bool) []int {
	var result []int
	for _, n := range nums {
		if test(n) {
			result = append(result, n)
		}
	}
	return result
}

func square(n int) int {
    return n*n
}
func mapInts(nums []int, transform func(int) int) []int {
	var result []int
	for _, n := range nums {
		result = append(result, transform(n))
	}
	return result
}

//  functional

func isEven(n int) bool {
	return n%2 == 0
}



func process(nums []int) []int {
	var result []int
	for _, n := range nums {
		if isEven(n) {
			result = append(result, square(n))
		}
	}
	return result
}
